tests/integration/security/reachability_test.go

```diff
@@ -119,22 +127,38 @@ func TestReachability(t *testing.T) {
 					ExpectSuccess: func(src echo.Instance, opts echo.CallOptions) bool {
 						// autoMtls doesn't work for client that doesn't have proxy, unless target doesn't
 						// have proxy or have mTLS disabled
-						if src == rctx.Naked {
-							return opts.Target == rctx.Naked || (opts.Target == rctx.B && opts.PortName != "http")
+						if rctx.IsNaked(src) {
+							return rctx.IsNaked(opts.Target) || (opts.Target == rctx.B && opts.PortName != "http")
 
 						}
+						// headless with sidecar injected, global mTLS enabled, no client side transport socket or transport_socket_matches since it's headless service.
+						if src != rctx.Headless && opts.Target == rctx.Headless {
```

Comment:
and the trick here is that if the opts.Target == rctx.HeadlessNaked, we will return true, while the global mTLS is enabled. This is the different case that the changes in Pilot make it from fail to success.

I feel we can somehow make test this case's condition more obvious, but we have already spent quite some time on this PR. let's ship this first.